File: python-fastapi/app/fhir_mapper.py
"""Bidirectional mapper between SQLAlchemy models and FHIR R4 resources.

Uses the ``fhir.resources`` library which gives us Pydantic-based FHIR
models with built-in validation against the R4 spec.
"""

from __future__ import annotations

from datetime import date, datetime, timezone
from typing import Any

from app.models.patient import PatientModel
from app.models.coverage import CoverageModel
from app.models.claim import ClaimModel
from app.models.eob import ExplanationOfBenefitModel

SYSTEM_MRN = "http://hospital.example.org/mrn"
SYSTEM_CPT = "http://www.ama-assn.org/go/cpt"
SYSTEM_CLAIM_TYPE = "http://terminology.hl7.org/CodeSystem/claim-type"
SYSTEM_ADJUDICATION = "http://terminology.hl7.org/CodeSystem/adjudication"


# --- Patient ---

def patient_to_fhir(entity: PatientModel) -> dict[str, Any]:
    resource: dict[str, Any] = {
        "resourceType": "Patient",
        "id": str(entity.id),
        "identifier": [{"system": SYSTEM_MRN, "value": entity.mrn}],
        "name": [{"family": entity.family_name, "given": [entity.given_name]}],
        "meta": {"lastUpdated": _to_iso(entity.updated_at)},
    }
    if entity.birth_date:
        resource["birthDate"] = entity.birth_date.isoformat()
    if entity.gender:
        resource["gender"] = entity.gender
    telecom = []
    if entity.phone:
        telecom.append({"system": "phone", "value": entity.phone})
    if entity.email:
        telecom.append({"system": "email", "value": entity.email})
    if telecom:
        resource["telecom"] = telecom
    if entity.address_line:
        resource["address"] = [{
            "line": [entity.address_line],
            "city": entity.city,
            "state": entity.state,
            "postalCode": entity.postal_code,
        }]
    return resource


def fhir_to_patient_fields(fhir: dict[str, Any]) -> dict[str, Any]:
    fields: dict[str, Any] = {}
    if names := fhir.get("name"):
        name = names[0]
        fields["family_name"] = name.get("family", "")
        fields["given_name"] = " ".join(name.get("given", []))
    if identifiers := fhir.get("identifier"):
        fields["mrn"] = identifiers[0].get("value", "")
    if bd := fhir.get("birthDate"):
        fields["birth_date"] = date.fromisoformat(bd)
    if g := fhir.get("gender"):
        fields["gender"] = g
    return fields


# --- Coverage ---

def coverage_to_fhir(entity: CoverageModel) -> dict[str, Any]:
    resource: dict[str, Any] = {
        "resourceType": "Coverage",
        "id": str(entity.id),
        "status": entity.status,
        "beneficiary": {"reference": f"Patient/{entity.patient_id}"},
        "subscriberId": entity.subscriber_id,
        "payor": [{"reference": f"Organization/{entity.payor_id}", "display": entity.payor_name}],
        "relationship": {"coding": [{"code": entity.relationship}]},
    }
    if entity.period_start:
        period: dict[str, str] = {"start": entity.period_start.isoformat()}
        if entity.period_end:
            period["end"] = entity.period_end.isoformat()
        resource["period"] = period
    return resource


# --- Claim ---

def claim_to_fhir(entity: ClaimModel) -> dict[str, Any]:
    items = []
    for item in entity.items:
        items.append({
            "sequence": item.sequence,
            "productOrService": {
                "coding": [{
                    "system": SYSTEM_CPT,
                    "code": item.product_or_service_code,
                    "display": item.product_or_service_display,
                }]
            },
            "quantity": {"value": item.quantity},
            "unitPrice": {"value": float(item.unit_price), "currency": entity.currency},
            "net": {"value": float(item.net_amount), "currency": entity.currency},
        })

    resource: dict[str, Any] = {
        "resourceType": "Claim",
        "id": str(entity.id),
        "status": entity.status,
        "type": {"coding": [{"system": SYSTEM_CLAIM_TYPE, "code": entity.type}]},
        "use": entity.use,
        "patient": {"reference": f"Patient/{entity.patient_id}"},
        "insurance": [{
            "sequence": 1,
            "focal": True,
            "coverage": {"reference": f"Coverage/{entity.coverage_id}"},
        }],
        "billablePeriod": {
            "start": entity.billable_period_start.isoformat(),
            "end": entity.billable_period_end.isoformat(),
        },
        "total": {"value": float(entity.total_amount), "currency": entity.currency},
        "item": items,
        "meta": {"lastUpdated": _to_iso(entity.updated_at)},
    }
    if entity.provider_reference:
        resource["provider"] = {"reference": entity.provider_reference}
    return resource


# --- ExplanationOfBenefit ---

def eob_to_fhir(entity: ExplanationOfBenefitModel) -> dict[str, Any]:
    resource: dict[str, Any] = {
        "resourceType": "ExplanationOfBenefit",
        "id": str(entity.id),
        "status": entity.status,
        "outcome": entity.outcome,
        "type": {"coding": [{"system": SYSTEM_CLAIM_TYPE, "code": entity.type}]},
        "patient": {"reference": f"Patient/{entity.patient_id}"},
        "claim": {"reference": f"Claim/{entity.claim_id}"},
        "meta": {"lastUpdated": _to_iso(entity.updated_at)},
    }
    if entity.disposition:
        resource["disposition"] = entity.disposition

    total = []
    if entity.total_submitted is not None:
        total.append({
            "category": {"coding": [{"system": SYSTEM_ADJUDICATION, "code": "submitted"}]},
            "amount": {"value": float(entity.total_submitted), "currency": "USD"},
        })
    if entity.total_benefit is not None:
        total.append({
            "category": {"coding": [{"system": SYSTEM_ADJUDICATION, "code": "benefit"}]},
            "amount": {"value": float(entity.total_benefit), "currency": "USD"},
        })
    if total:
        resource["total"] = total

    if entity.total_benefit is not None:
        resource["payment"] = {
            "amount": {"value": float(entity.total_benefit), "currency": "USD"}
        }
    return resource


# --- Bundle helper ---

def search_bundle(resources: list[dict[str, Any]]) -> dict[str, Any]:
    entries = []
    for r in resources:
        rt = r.get("resourceType", "Resource")
        rid = r.get("id", "")
        entries.append({"fullUrl": f"{rt}/{rid}", "resource": r})
    return {
        "resourceType": "Bundle",
        "type": "searchset",
        "total": len(resources),
        "entry": entries,
    }


def _to_iso(dt: datetime | None) -> str | None:
    if dt is None:
        return None
    if dt.tzinfo is None:
        dt = dt.replace(tzinfo=timezone.utc)
    return dt.isoformat()

File: node-express/src/models/fhirMapper.js
const SYSTEM_MRN = 'http://hospital.example.org/mrn';
const SYSTEM_CPT = 'http://www.ama-assn.org/go/cpt';
const SYSTEM_CLAIM_TYPE = 'http://terminology.hl7.org/CodeSystem/claim-type';
const SYSTEM_ADJUDICATION = 'http://terminology.hl7.org/CodeSystem/adjudication';

// --- Patient ---

function patientToFhir(row) {
  const resource = {
    resourceType: 'Patient',
    id: row.id,
    identifier: [{ system: SYSTEM_MRN, value: row.mrn }],
    name: [{ family: row.family_name, given: [row.given_name] }],
    meta: { lastUpdated: row.updated_at },
  };
  if (row.birth_date) resource.birthDate = row.birth_date;
  if (row.gender) resource.gender = row.gender;
  const telecom = [];
  if (row.phone) telecom.push({ system: 'phone', value: row.phone });
  if (row.email) telecom.push({ system: 'email', value: row.email });
  if (telecom.length) resource.telecom = telecom;
  if (row.address_line) {
    resource.address = [{
      line: [row.address_line], city: row.city,
      state: row.state, postalCode: row.postal_code,
    }];
  }
  return resource;
}

function fhirToPatientFields(body) {
  const fields = {};
  if (body.name?.[0]) {
    fields.family_name = body.name[0].family || '';
    fields.given_name = (body.name[0].given || []).join(' ');
  }
  if (body.identifier?.[0]) fields.mrn = body.identifier[0].value;
  if (body.birthDate) fields.birth_date = body.birthDate;
  if (body.gender) fields.gender = body.gender;
  return fields;
}

// --- Coverage ---

function coverageToFhir(row) {
  const resource = {
    resourceType: 'Coverage',
    id: row.id,
    status: row.status,
    beneficiary: { reference: `Patient/${row.patient_id}` },
    subscriberId: row.subscriber_id,
    payor: [{ reference: `Organization/${row.payor_id}`, display: row.payor_name }],
    relationship: { coding: [{ code: row.relationship }] },
  };
  if (row.period_start) {
    resource.period = { start: row.period_start };
    if (row.period_end) resource.period.end = row.period_end;
  }
  return resource;
}

// --- Claim ---

function claimToFhir(row, items = []) {
  const resource = {
    resourceType: 'Claim',
    id: row.id,
    status: row.status,
    type: { coding: [{ system: SYSTEM_CLAIM_TYPE, code: row.type }] },
    use: row.use,
    patient: { reference: `Patient/${row.patient_id}` },
    insurance: [{ sequence: 1, focal: true, coverage: { reference: `Coverage/${row.coverage_id}` } }],
    billablePeriod: { start: row.billable_period_start, end: row.billable_period_end },
    total: { value: parseFloat(row.total_amount), currency: row.currency },
    item: items.map(i => ({
      sequence: i.sequence,
      productOrService: {
        coding: [{ system: SYSTEM_CPT, code: i.product_or_service_code, display: i.product_or_service_display }],
      },
      quantity: { value: i.quantity },
      unitPrice: { value: parseFloat(i.unit_price), currency: row.currency },
      net: { value: parseFloat(i.net_amount), currency: row.currency },
    })),
    meta: { lastUpdated: row.updated_at },
  };
  if (row.provider_reference) resource.provider = { reference: row.provider_reference };
  return resource;
}

// --- ExplanationOfBenefit ---

function eobToFhir(row) {
  const resource = {
    resourceType: 'ExplanationOfBenefit',
    id: row.id,
    status: row.status,
    outcome: row.outcome,
    type: { coding: [{ system: SYSTEM_CLAIM_TYPE, code: row.type }] },
    patient: { reference: `Patient/${row.patient_id}` },
    claim: { reference: `Claim/${row.claim_id}` },
    meta: { lastUpdated: row.updated_at },
  };
  if (row.disposition) resource.disposition = row.disposition;
  const total = [];
  if (row.total_submitted != null) {
    total.push({
      category: { coding: [{ system: SYSTEM_ADJUDICATION, code: 'submitted' }] },
      amount: { value: parseFloat(row.total_submitted), currency: 'USD' },
    });
  }
  if (row.total_benefit != null) {
    total.push({
      category: { coding: [{ system: SYSTEM_ADJUDICATION, code: 'benefit' }] },
      amount: { value: parseFloat(row.total_benefit), currency: 'USD' },
    });
  }
  if (total.length) resource.total = total;
  if (row.total_benefit != null) {
    resource.payment = { amount: { value: parseFloat(row.total_benefit), currency: 'USD' } };
  }
  return resource;
}

// --- Bundle ---

function searchBundle(resources) {
  return {
    resourceType: 'Bundle',
    type: 'searchset',
    total: resources.length,
    entry: resources.map(r => ({
      fullUrl: `${r.resourceType}/${r.id}`,
      resource: r,
    })),
  };
}

module.exports = {
  patientToFhir, fhirToPatientFields,
  coverageToFhir, claimToFhir, eobToFhir, searchBundle,
};
